core/state_processor.go

```diff
@@ -869,81 +927,103 @@ func (p *ParallelStateProcessor) Process(block *types.Block, statedb *state.Stat
+			baseStateDB:     statedb,
+			staticSlotIndex: -1,
+			tx:              tx,
+			gasLimit:        block.GasLimit(), // gp.Gas().
+			msg:             msg,
+			block:           block,
+			vmConfig:        cfg,
+			bloomProcessor:  bloomProcessor,
+			usedGas:         usedGas,
+			curTxChan:       make(chan int, 1),
+			systemAddrRedo:  false, // set to true, when systemAddr access is detected.
+			runnable:        1,     // 0: not runnable, 1: runnable
+			executedNum:     0,
 		}
+		p.allTxReqs = append(p.allTxReqs, txReq)
+	}
+	// set up stage2 enter criteria
+	p.targetStage2Count = len(p.allTxReqs)
+	if p.targetStage2Count > 50 {
+		// usually, the the last Tx could be the bottleneck it could be very slow,
+		// so it is better for us to enter stage 2 a bit earlier
+		p.targetStage2Count = p.targetStage2Count - stage2AheadNum
+	}
 
-		// to optimize the for { for {} } loop code style? it is ok right now.
-		for {
-			if p.queueSameFromAddress(txReq) {
-				break
-			}
+	p.doStaticDispatch(statedb, p.allTxReqs) // todo: put txReqs in unit?
+	// after static dispatch, we notify the slot to work.
+	for _, slot := range p.slotState {
+		slot.primaryWakeUpChan <- struct{}{}
+	}
+	// wait until all Txs have processed.
+	for {
+		if len(commonTxs)+len(systemTxs) == txNum {
+			// put it ahead of chan receive to avoid waiting for empty block
+			break
+		}
+
+		unconfirmedResult := <-p.txResultChan
+		unconfirmedTxIndex := unconfirmedResult.txReq.txIndex
+		if unconfirmedTxIndex <= p.mergedTxIndex {
+			// log.Warn("drop merged txReq", "unconfirmedTxIndex", unconfirmedTxIndex, "p.mergedTxIndex", p.mergedTxIndex)
+			continue
+		}
+		p.pendingConfirmResults[unconfirmedTxIndex] = append(p.pendingConfirmResults[unconfirmedTxIndex], unconfirmedResult)
+
+		// schedule prefetch once only when unconfirmedResult is valid
+		if unconfirmedResult.err == nil {
+			if _, ok := p.txReqExecuteRecord[unconfirmedTxIndex]; !ok {
+				p.txReqExecuteRecord[unconfirmedTxIndex] = 0
+				p.txReqExecuteCount++
+				statedb.AddrPrefetch(unconfirmedResult.slotDB) // todo: prefetch when it is not merged
+				// enter stage2, RT confirm
+				if !p.inConfirmStage2 && p.txReqExecuteCount == p.targetStage2Count {
+					p.inConfirmStage2 = true
+				}
 
-			if p.queueSameToAddress(txReq) {
-				break
 			}
-			// if idle slot available, just dispatch and process next tx.
-			if p.dispatchToIdleSlot(statedb, txReq) {
+			p.txReqExecuteRecord[unconfirmedTxIndex]++
+		}
+
+		for {
+			result := p.confirmTxResults(statedb, gp)
+			if result == nil {
 				break
 			}
-			log.Debug("ProcessParallel no slot available, wait", "txIndex", txReq.txIndex)
-			// no idle slot, wait until a tx is executed and merged.
-			result := p.waitUntilNextTxDone(statedb, gp)
-
 			// update tx result
 			if result.err != nil {
-				log.Warn("ProcessParallel a failed tx", "resultSlotIndex", result.slotIndex,
+				log.Error("ProcessParallel a failed tx", "resultSlotIndex", result.slotIndex,
```

Comment:
Do we need to do clean up here before return

Reply:
done, thx